### ISSUE TITLE
Add reportDisableDirectives config/CLI option

### ISSUE THREAD

**Issue by galvarez421** [eslint/eslint#12304]
**The version of ESLint you are using.**
v6.3.0

**The problem you want to solve.**
As far as I can tell, it is not possible to configure ESLint to report warnings or errors for `eslint-disable*` comments in such a way that wouldn't allow for these types of comments to disable those warnings or errors -- at least not without some side effects, as describe below.

My use case is that I would like for an ESLint config to be able to discourage ('warn') or disallow ('error') the use of ESLint disable directives/comments.

A rule like `'no-warning-comments': [ 'warn', { terms: [ 'eslint-disable' ], } ]` or even a new rule like `'no-disable-comments': 'error'`doesn't exactly satisfy the need because a user can disable any rule using an `eslint-disable*` comment. Therefore, if the configuration were only possible via a rule, then it would never actually report a `/* eslint-disable */` comment (with the exception described below), for example, rendering the rule ineffectual.

The configuration option `noInlineConfig: true` doesn't exactly satisfy the need either because while it does prevent `eslint-disable*` comments from disabling warnings or errors for a rule disallowing `eslint-disable*` comments (such as the rules mentioned above), it also renders `eslint-disable*` comments ineffectual (not the goal of the proposal) and renders all inline configuration comments ineffectual (also not the goal of the proposal).

**Your take on the correct solution to problem.**
Perhaps a top-level configuration/CLI option (in the vein of `noInlineConfig` and `reportUnusedDisableDirectives`) could be added that would report on the usage of `eslint-disable*` comments. For example, an option could be added that would be called `reportDisableDirectives`. Since the desired behavior of the option is essentially that of a rule, it should probably be configurable like a standard rule, e.g., `reportDisableDirectives: 'warn'` or `reportDisableDirectives: 'error'` or `reportDisableDirectives: [ 'warn', { someOption: true } ]`, etc. Importantly, it should not be possible to disable this option from reporting violations using disable comments, since that would defeat the purpose of the option (as described previously).

**Are you willing to submit a pull request to implement this change?**
No.

**Comment by g-plane** [comment]
Thanks for this new issue.

There is an [ESLint plugin](https://github.com/mysticatea/eslint-plugin-eslint-comments) which is related to ESLint comments and you may take a look. However, I'm not sure if that plugin can satisfy your request.

**Comment by kaicataldo** [comment]
While it's great to get feedback here, we ask that all changes to core go through the RFC process outlined [here](https://github.com/eslint/rfcs).

Having tried to solve this myself in previous jobs, my initial take is that at some point there has to be some level of trust of contributors and some sort of system of manual review that will catch things like this. Completely disallowing disable comments means that CI could potentially be blocked on every semver-minor release of ESLint (since they can increase errors according to [our policy](https://github.com/eslint/eslint/#semantic-versioning-policy)).

**Comment by galvarez421** [comment]
@g-plane Thank you for the information. I am in the process of determining if that plugin satisfies my use case. It looks promising.

@kaicataldo Thank you for the information about the RFC process. It seems that process applies more for people intending to submit a pull request right? I don't think I understand your CI example. In that example, if that scenario is considered unacceptable for a project, wouldn't it make sense either not to use the proposed feature or to configure it only to warn?

**Comment by kaicataldo** [comment]
> It seems that process applies more for people intending to submit a pull request right?

Unless I'm misunderstanding the proposal, I think this would apply any time the results of running ESLint could block a process.

For what it's worth, I've tried to use [no-restricted-disable](https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/no-restricted-disable.html) from the plugin suggested above in the past, but it still requires manual review because it's possible to disable the rule itself with another disable comment (as mentioned in the original comment).

If we were to implement something like this, I wonder if it makes more sense for it to be a boolean flag rather than configurable as `"warn"` or `"error"`. I don't think it should be treated the same as rules since it's a configuration flag, and if a project disallows using disable comments, it feels like it should error on them. A plugin like `eslint-plugin-eslint-comments` could still be used to report with a warning level, if that's the desired behavior.

**Comment by galvarez421** [comment]
Thank you for the feedback @kaicataldo.

> I wonder if it makes more sense for it to be a boolean flag rather than configurable as "warn" or "error". I don't think it should be treated the same as rules since it's a configuration flag

Essentially what I am requesting is a rule, except the rule would inherently require being treated differently than other rules otherwise the very thing that is supposed to be reported by the rule could disable the rule and thus the report (I'm not sure there is another conceivable rule that would share this trait). That is the only reason I suggested a top-level configuration option rather than a rule that would behave differently with regard to disable comments (if such a thing were even possible). I also didn't mention this in the OP but I think it would make sense for such an option/rule to account for `/* eslint {rule-name}: off */` comments as well since those are functionally equivalent to `eslint-disable*` comments as far as I can tell.

> if a project disallows using disable comments, it feels like it should error on them. A plugin like eslint-plugin-eslint-comments could still be used to report with a warning level, if that's the desired behavior.

For that matter, that plugin could be used to report errors as well right? From reading the documentation, I think the only way in which the plugin's `no-use` rule does not satisfy the use case is that it still allows for users to disable the rule via `/* eslint {rule-name}: off */` comments, per https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/no-use.html#known-limitations.
